architecture_records/002-acceleration-framework.md

```diff
@@ -0,0 +1,480 @@
+# Training Enhancements Framework
+
+**Deciders(s)**: Sukriti Sharma (sukriti.sharma4@ibm.com), Raghu Ganti (rganti@us.ibm.com), Laura Wynter (lwynter@sg.ibm.com), Fabian Lim (flim@sg.ibm.com), Aaron Chew (aaron.chew1@ibm.com)
+**Date (YYYY-MM-DD)**:  2024-04-11
+**Obsoletes ADRs**:  NA
+**Modified By ADRs**:  NA
+**Relevant Issues**: [116](https://github.com/foundation-model-stack/fms-hf-tuning/pull/116)
+
+- [Summary and Objective](#summary-and-objective)
+  - [Motivation](#motivation)
+  - [User Benefit](#user-benefit)
+- [Decision](#decision)
+  - [Alternatives Considered](#alternatives-considered)
+- [Consequences](#consequences)
+- [Detailed Design](#detailed-design)
+
+## Summary and Objective
+
+Design and implement a framework to include custom acceleration tools into [`sft_trainer.py`], that improve training metrics such as GPU memory consumption, training speed, etc.
+
+<!--
+Context goes here.
+
+Describe the forces at play, including technological, political, social, and project local. These forces are likely in tension, and should be called out as such. The language in this section is value-neutral. It is simply describing facts.
+-->
+
+### Motivation
+
+Currently `sft_trainer.py` only can access those tools already integrated in HF. Due to rapid developments in AI training methodologies, these technologies are quickly considered "standard", for example:
+1. LoRA adapters from [PEFT](https://github.com/huggingface/peft).
+2. Prefix tuning from [PEFT](https://github.com/huggingface/peft).
+3. FSDP training from [accelerate](https://github.com/huggingface/accelerate).
```

Comment:
These are examples of features that _are_ part of HF training. I'm not clear if the intent is to say that these are features that could have been implemented outside HF and still used in HF trainer? If so I think it should be a bit more explicit

Reply:
Sorry for the confusion! The intent was not to say that innovations come very quickly, and while these 3 already have some form of integration in HF, and thus are readily accessible, there are others that are currently not. I have listed those  a couple of lines below. 

> Recently, it has been observed that new training techniques are released with an incomplete "preview" version. These "preview" versions tend to be not be fully integrated into OSS. Therefore, using new techniques typically involve additional work. This framework aims to allow timely integrations of such techniques into `sft_trainer.py`. A short list of powerful training techniques but are "preview"-only include:
> - Huggingface integration of [AutoGPTQ](https://github.com/AutoGPTQ/AutoGPTQ).
    * 4-bit quantization kernels to reduce memory storage of the base weights.
> - [Unsloth](https://github.com/unslothai/unsloth).
    * Fused operation kernels
    * Kernels for common model architectures (e.g., cross-entropy losses, RoPE embeddings and RMS norms).
> - [megablocks](https://github.com/databricks/megablocks).
    * acceleration package for distributing mixture-of-experts that improves upon FSDP sharding.

The concern of this ADR is then, how to get these advancements (any many more in the future) working with `tuning.sft_trainer.py` in a timely fashion, where we can integrate, fix issues and test, without having to make a PRs to constantly change , and most invariably bloat, `tuning.sft_trainer.py`. 

I believe having something like this will greatly increase the value of `tuning.sft_trainer.py`.

Reply:
> The intent was not to say that innovations come very quickly

"not" looks like a typo?

I think what I'm suggesting is just a slight tweak to the wording of the doc to be more explicit/clear about what you're saying this feature will/won't be able to do. (not for my sake, but for the sake of readers in general).

Since you don't say "these are techniques that could be / could have been implemented with this plugin framework" this can be interpreted as _some_ could have been ... or since you say "aims to" and "such techniques" can be interpreted as "some similar ones maybe, but not necessarily these specific ones at all and not really sure."

Also do we have examples of any yet-unimplemented features we would hope to use this for?

Reply:
Hi @nairbv . Sorry for the typo in my [comment above](https://github.com/foundation-model-stack/fms-hf-tuning/pull/119#discussion_r1580240177). I have since corrected it. 

Also I have attempted to address your concern in my latest update.